buildSrc/src/main/groovy/net/minecraftforge/forge/tasks/checks/CheckATs.groovy

```diff
@@ -0,0 +1,304 @@
+                    } else {
+                        jcls.fields.each { field, value ->
+                            final fkey = entry.cls + ' ' + field
+                            if (accessLevel(value.access) < accessStr(entry.modifier)) {
+                                if (lines.containsKey(fkey)) {
+                                    toRemove.add(fkey)
+                                } else if (!entry.existing.contains(fkey)) {
+                                    println('Added: ' + fkey)
+                                }
+                                entry.children.add(fkey)
+                            } else if (lines.containsKey(fkey)) {
+                                toRemove.add(fkey)
+                                println('Removed: ' + fkey)
+                            }
+                        }
+                        entry.existing.findAll { it !in entry.children }.each { println('Removed: ' + it) }
+                    }
+                } else if ('*()' == entry.desc) {
+                    if (!jcls.methods) {
+                        itr.remove()
+                        reporter.report("Invalid group, class has no methods: $key")
+                    } else {
+                        jcls.methods.each { mtd, value ->
+                            if (mtd.startsWith('<clinit>') || mtd.startsWith('lambda$')) return
+                            key = entry.cls + ' ' + mtd.replace(' ', '')
+                            if (accessLevel(value.access) < accessStr(entry.modifier)) {
+                                if (lines.containsKey(key)) {
+                                    toRemove.add(key)
+                                } else if (!entry.existing.contains(key)) {
+                                    println('Added: ' + key)
+                                }
+                                entry.children.add(key)
+                            } else if (lines.containsKey(key)) {
+                                toRemove.add(key)
+                                println('Removed: ' + key)
+                            }
+                        }
+                        entry.existing.findAll { it !in entry.children }.each { println('Removed: ' + it) }
+                    }
+                } else if ('<init>' == entry.desc) { //Make all public non-abstract subclasses
+                    inheritance.each { tcls, value ->
+                        if (!value.methods || ((value.access & Opcodes.ACC_ABSTRACT) != 0)) return
+                        final parents = [] as Set
+                        String parent = tcls
+                        while (parent !== null && inheritance.containsKey(parent)) {
+                            parents.add(parent)
+                            def p = inheritance[parent]
+                            parent = p === null ? null : p.superName
+                        }
+                        if (parents.contains(entry.cls.replaceAll('\\.', '/'))) {
```

Comment:
Do we really need to traverse all the parents of the target class _then_ check if the AT class (`entry.cls`) is a parent? Could we not do the check at each parent name, and short-circuit once we encounter the parent for the first time? This would also avoid allocating a `Set` each time.